evennia/objects/manager.py

```diff
@@ -486,51 +440,59 @@ def _search_by_tag(query, taglist):
+                else:
+                    typeclasses[i] = "%s" % typeclass
+            typeclass = typeclasses
+
         # Search through all possibilities.
         match_number = None
         # always run first check exact - we don't want partial matches
         # if on the form of 1-keyword etc.
         matches = _searcher(searchdata, candidates, typeclass, exact=True)
 
+        stripped_searchdata = searchdata
         if not matches:
             # no matches found - check if we are dealing with N-keyword
             # query - if so, strip it.
-            match = _MULTIMATCH_REGEX.match(str(searchdata))
+            match_data = _MULTIMATCH_REGEX.match(str(searchdata))
             match_number = None
-            stripped_searchdata = searchdata
-            if match:
+            if match_data:
                 # strips the number
-                match_number, stripped_searchdata = match.group("number"), match.group("name")
+                match_number, stripped_searchdata = match_data.group("number"), match_data.group(
+                    "name"
+                )
                 match_number = int(match_number) - 1
             if match_number is not None:
                 # run search against the stripped data
                 matches = _searcher(stripped_searchdata, candidates, typeclass, exact=True)
-                if not matches:
-                    # final chance to get a looser match against the number-strippped query
-                    matches = _searcher(stripped_searchdata, candidates, typeclass, exact=False)
-            elif not exact:
-                matches = _searcher(searchdata, candidates, typeclass, exact=False)
 
-        if tags:
-            matches = _search_by_tag(matches, make_iter(tags))
+        # at this point, if there are no matches, we give it a chance to find fuzzy matches
+        if not exact and not matches:
+            # we use stripped_searchdata in case a match number was included
+            matches = _searcher(stripped_searchdata, candidates, typeclass, exact=False)
 
         # deal with result
-        if len(matches) == 1 and match_number is not None and match_number != 0:
-            # this indicates trying to get a single match with a match-number
-            # targeting some higher-number match (like 2-box when there is only
-            # one box in the room). This leads to a no-match.
-            matches = self.none()
-        elif len(matches) > 1 and match_number is not None:
-            # multiple matches, but a number was given to separate them
-            if 0 <= match_number < len(matches):
+        if match_number is not None:
+            # this indicates someone attempting to get a single match with a match-number
+            if match_number >= len(matches):
+                # the search attempted to target a higher match index than exists in the candidates.
+                # This leads to a no-match.
+                matches = self.none()
```

Comment:
This first `if` conditional seems like it is superfluous ... both this and the `else:` statement fulfills the same condition (being outside of the range).

Reply:
This still seems to be the case.

Reply:
The outer `if match_number is not None` is functionally relevant. In terms of code logic, it means that the inner checks regarding whether the match-number index is valid will only be checked if there is, in fact, a match-number provided in the term to begin with. I don't believe this to be superfluous or redundant.

Reply:
Ah, I've been misinterpreting what you meant by "the first if"; since it's below two ifs I thought you meant the first of the two you commended below.